### PR TITLE
Training BEVFusion successfully.

### PR DESCRIPTION
**Pull request by daxiongpro**
## Motivation

Integrating BEV-Fusion into the MMDetection3D framework.

## Modification

Fix some bug and can train successfully.
Wrote a script for one-click environment installation in `tools/env_install.sh` file.

## BC-breaking (Optional)

Please follow the version as:
mmcv==2.0.0rc4
mmdet==3.0.0rc6
mmdet3d==1.0.0rc3
mmengine==0.6.0

## Use cases (Optional)

`python tools/train.py projects/BEVFusion/configs/bevfusion_voxel0075_second_secfpn_8xb4-cyclic-20e_nus-3d.py`

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
@daxiongpro Thanks for your PR. Please sign the CLA. Also, how about the precision of the trained mode?

**Comment by daxiongpro** [comment]
> @daxiongpro Thanks for your PR. Please sign the CLA. Also, how about the precision of the trained mode?

Hi，I met a strange problem. When I train the code after 6 epoch, the job will stop:
`03/06 13:58:55 - mmengine - INFO - Epoch(train)  [6][14050/14065]  lr: 1.3584e-05  eta: 2 days, 2:30:46  time: 0.9324  data_time: 0.0141  memory: 12152  grad_norm: 15.8058  loss: 2.9092  loss_heatmap: 1.2752  layer_-1_loss_cls: 0.3059  layer_-1_loss_bbox: 1.3281  matched_ious: 0.2958
03/06 13:59:08 - mmengine - INFO - Exp name: bevfusion_voxel0075_second_secfpn_8xb4-cyclic-20e_nus-3d_20230306_102014
03/06 13:59:08 - mmengine - INFO - Saving checkpoint at 6 epochs
/home/daxiongpro/code/mmdetection3d/projects/BEVFusion/bevfusion/transfusion_head.py:256: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  top_proposals_class = top_proposals // heatmap.shape[-1]
/home/daxiongpro/code/mmdetection3d/projects/BEVFusion/bevfusion/transfusion_head.py:685: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  feature_map_size = (grid_size[:2] // self.train_cfg['out_size_factor']
Traceback (most recent call last):
  File "tools/train.py", line 135, in <module>
    main()
  File "tools/train.py", line 131, in main
    runner.train()
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/mmengine/runner/runner.py", line 1686, in train
    model = self.train_loop.run()  # type: ignore
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/mmengine/runner/loops.py", line 90, in run
    self.run_epoch()
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/mmengine/runner/loops.py", line 106, in run_epoch
    self.run_iter(idx, data_batch)
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/mmengine/runner/loops.py", line 122, in run_iter
    outputs = self.runner.model.train_step(
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/mmengine/model/base_model/base_model.py", line 114, in train_step
    losses = self._run_forward(data, mode='loss')  # type: ignore
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/mmengine/model/base_model/base_model.py", line 326, in _run_forward
    results = self(**data, mode=mode)
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/daxiongpro/code/mmdetection3d/mmdet3d/models/detectors/base.py", line 75, in forward
    return self.loss(inputs, data_samples, **kwargs)
  File "/home/daxiongpro/code/mmdetection3d/projects/BEVFusion/bevfusion/bevfusion.py", line 253, in loss
    loss = self.bbox_head.loss(gt_bboxes_3d, gt_labels_3d, preds_dicts, **kwargs)
  File "/home/daxiongpro/code/mmdetection3d/projects/BEVFusion/bevfusion/transfusion_head.py", line 749, in loss
    ) = self.get_targets(gt_bboxes_3d, gt_labels_3d, preds_dicts[0])
  File "/home/daxiongpro/code/mmdetection3d/projects/BEVFusion/bevfusion/transfusion_head.py", line 527, in get_targets
    res_tuple = multi_apply(
  File "/home/daxiongpro/.conda/envs/open-mmlab-bevfusion/lib/python3.8/site-packages/mmdet/models/utils/misc.py", line 219, in multi_apply
    return tuple(map(list, zip(*map_results)))
  File "/home/daxiongpro/code/mmdetection3d/projects/BEVFusion/bevfusion/transfusion_head.py", line 604, in get_targets_single
    assign_result = self.bbox_assigner.assign(
  File "/home/daxiongpro/code/mmdetection3d/projects/BEVFusion/bevfusion/utils.py", line 293, in assign
    matched_row_inds, matched_col_inds = linear_sum_assignment(cost)
ValueError: matrix contains invalid numeric entries
`

And I use the epoch_5 pth file to test, the result is:
`Evaluating bboxes of pred_instances_3d
mAP: 0.4009
mATE: 0.4187
mASE: 0.3755
mAOE: 0.8531
mAVE: 1.1477
mAAE: 0.2666
NDS: 0.4091
Eval time: 119.0s

Per-class results:
Object Class    AP      ATE     ASE     AOE     AVE     AAE
car     0.784   0.209   0.241   0.389   0.746   0.238
truck   0.325   0.531   0.387   0.589   1.179   0.353
bus     0.412   0.537   0.326   0.745   2.719   0.452
trailer 0.112   0.795   0.395   1.473   0.753   0.268
construction_vehicle    0.065   0.833   0.537   1.486   0.149   0.376
pedestrian      0.770   0.174   0.295   0.686   0.503   0.168
motorcycle      0.336   0.280   0.361   0.764   2.366   0.195
bicycle 0.126   0.251   0.411   1.270   0.766   0.082
traffic_cone    0.543   0.203   0.415   nan     nan     nan
barrier 0.537   0.374   0.387   0.275   nan     nan`

**Comment by daxiongpro** [comment]
> 

I've solved the problem.

**Comment by daxiongpro** [comment]
I find the loss exploded after 6 epoch.

File: projects/BEVFusion/configs/bevfusion_voxel0075_second_secfpn_8xb4-cyclic-20e_nus-3d.py [hunk 1]
```diff
@@ -246,12 +248,16 @@
...
+    #     is_train=True),
     dict(
         type='GlobalRotScaleTrans',
-        resize_lim=[0.9, 1.1],
-        rot_lim=[-0.78539816, 0.78539816],
-        trans_lim=0.5,
-        is_train=True),
+        rot_range=[-0.78539816, 0.78539816],
+        scale_ratio_range=[0.9, 1.1]),
     dict(type='RandomFlip3D'),
```
Comment (JingweiZhang12) [inline projects/BEVFusion/configs/bevfusion_voxel0075_second_secfpn_8xb4-cyclic-20e_nus-3d.py:261]: I noticed that the implementation of RandomFlip3D in mmdet3d and BEVFusion is different. You can check it.

File: projects/BEVFusion/bevfusion/bevfusion.py [hunk 2]
```diff
@@ -239,4 +239,19 @@ def extract_feat(
     def loss(self, batch_inputs_dict: Dict[str, Optional[Tensor]],
              batch_data_samples: List[Det3DDataSample],
              **kwargs) -> List[Det3DDataSample]:
-        pass
+        batch_input_metas = [item.metainfo for item in batch_data_samples]
+        feats = self.extract_feat(batch_inputs_dict, batch_input_metas)
```
Comment (JingweiZhang12) [inline projects/BEVFusion/bevfusion/bevfusion.py:243]: The `GlobalRotScaleTrans` and `RandomFlip3D` in mmdet have not recorded the `lidar_aug_matrix` and `img_aug_matrix` like BEVFusion. We should use these aug matrices in image view transformation in `extrack_img_feat` function.

File: projects/BEVFusion/bevfusion/utils.py [hunk 3]
```diff
@@ -274,13 +275,19 @@ def assign(self, bboxes, gt_bboxes, gt_labels, cls_pred, train_cfg):
 
         # 2. compute the weighted costs
         # see mmdetection/mmdet/core/bbox/match_costs/match_cost.py
-        cls_cost = self.cls_cost(cls_pred[0].T, gt_labels)
+        pred_instances = InstanceData(scores=cls_pred[0].T)
+        gt_instances = InstanceData(labels=gt_labels)
+        cls_cost = self.cls_cost(pred_instances, gt_instances)
         reg_cost = self.reg_cost(bboxes, gt_bboxes, train_cfg)
-        iou = self.iou_calculator(bboxes, gt_bboxes)
+        iou = self.iou_calculator(bboxes, gt_bboxes)  # (200, 4) 有nan
```
Comment (JingweiZhang12) [inline projects/BEVFusion/bevfusion/utils.py:282]: The `iou` in the training process in my implementation is normal.

File: tools/create_data.py [hunk 4]
```diff
@@ -293,14 +293,14 @@ def semantickitti_data_prep(info_prefix, out_dir):
...
             max_sweeps=args.max_sweeps)
-        test_version = f'{args.version}-test'
-        nuscenes_data_prep(
-            root_path=args.root_path,
-            info_prefix=args.extra_tag,
-            version=test_version,
-            dataset_name='NuScenesDataset',
-            out_dir=args.out_dir,
-            max_sweeps=args.max_sweeps)
+        # test_version = f'{args.version}-test'
```
Comment (JingweiZhang12) [inline tools/create_data.py:296]: Please remove these unnecessary modifications.

File: projects/BEVFusion/configs/bevfusion_voxel0075_second_secfpn_8xb4-cyclic-20e_nus-3d.py [hunk 5]
```diff
@@ -221,17 +221,19 @@
         coord_type='LIDAR',
         load_dim=5,
         use_dim=5,
-        reduce_beams=32,
-        load_augmented=None),
+        # reduce_beams=32,
+        # load_augmented=None
+    ),
```
Comment (JingweiZhang12) [inline projects/BEVFusion/configs/bevfusion_voxel0075_second_secfpn_8xb4-cyclic-20e_nus-3d.py:226]: The `CBGS` setting should be used in training.

**Comment by JingweiZhang12** [comment]
@daxiongpro Hi, many thanks for your contiguous commits. Here is a training implementation of ours: https://github.com/JingweiZhang12/mmdetection3d/tree/bevfusion_train. It has some potential bugs but I have no enough time to fix them recently.  You can refer to it.  For the convenience of the subsequent communication, maybe we can add WeChat: zjw209624.

**Comment by daxiongpro** [comment]
Thanks for the author's reply. I found that when I set the batchsize to 40 (batchsize=8, 6 A100GPUs), the loss explosion problem disappeared.

**Comment by Monoclinic** [comment]
> @daxiongpro Hi, many thanks for your contiguous commits. Here is a training implementation of ours: https://github.com/JingweiZhang12/mmdetection3d/tree/bevfusion_train. It has some potential bugs but I have no enough time to fix them recently. You can refer to it. For the convenience of the subsequent communication, maybe we can add WeChat.

Hello, I found that you have successfully reproduced the result of BEVFusion in your implementation. I wonder if current version is reliable and whether it supports 1-stage training on lidar (15+5e).

File: projects/BEVFusion/bevfusion/utils.py [hunk 6]
```diff
@@ -279,11 +279,13 @@ def assign(self, bboxes, gt_bboxes, gt_labels, cls_pred, train_cfg):
...
         cls_cost = self.cls_cost(pred_instances, gt_instances)
         reg_cost = self.reg_cost(bboxes, gt_bboxes, train_cfg)
-        iou = self.iou_calculator(bboxes, gt_bboxes)
+        iou = self.iou_calculator(bboxes, gt_bboxes)  # (200, 4) 有nan
+        # iou = torch.where(torch.isnan(iou), torch.tensor([0.0]).cuda(), iou)  # nan 部分设置为 0
         iou_cost = self.iou_cost(iou)
 
         # weighted sum of above three costs
         cost = cls_cost + reg_cost + iou_cost
+        cost = torch.where(torch.isnan(cost), torch.tensor([0.0]).cuda(), cost)  # nan 部分设置为 0
```
Comment (ashrikant39) [inline projects/BEVFusion/bevfusion/utils.py]: I think you can use torch.nan_to_num.

**Comment by delightsomeness** [comment]
@daxiongpro 请问6代后梯度爆照的问题解决了嘛？我这里也遇到了同样的问题，除了增加batch_size 还有其他方法可以解决嘛？

**Comment by delightsomeness** [comment]
@daxiongpro 您好，请问您解决了训练BEVFusion时6代后梯度爆炸的问题了嘛？我这里也遇到了同样的问题。

**Pull request closed without merging**